model-optimizer/mo/ops/strided_slice.py

```diff
@@ -97,55 +54,86 @@ def convert(attr):
-                    i_port.data.set_value(new_value)
+        begin = node.in_port(1).data.get_value()
+        end = node.in_port(2).data.get_value()
+        if begin is None or end is None:
+            raise Error('StridedSlice operation supports only constant begin and end inputs')
+
+        if len(node.in_nodes()) > 3:
+            strides = node.in_port(3).data.get_value()
+            if strides is None:
+                raise Error('StridedSlice operation supports only constant strides input')
+        else:
+            strides = np.ones_like(begin)
+
+        shape = node.in_port(0).data.get_shape()
+        value = node.in_port(0).data.get_value()
+        input_rank = len(shape)
+        assert len(begin) == len(end) == len(strides), 'begin, end, and strides must be of the same length'
+
+        if shape is None or any([x < 0 for x in shape]):
+            return
+
+        # masks alligment all masks should initial slice_rank (mask_alligment != mask_extending)
+        extend_mask = lambda mask, val=0: np.append(mask, [val] * (len(begin) - len(mask))).astype(int)
+        new_axis_mask = extend_mask(node.new_axis_mask)
+        shrink_axis_mask = extend_mask(node.shrink_axis_mask)
+        begin_mask = extend_mask(node.begin_mask, 1)  # differs from the case when we unroll ellipsis
+        end_mask = extend_mask(node.end_mask, 1)  # when we unroll ellipsis zeros should be inserted
+        # no need to extend ellipsis
+
+        # unroll ellipsis
+        if np.any(node.ellipsis_mask):
+            i = np.nonzero(node.ellipsis_mask)
+            assert len(i[0]) == 1, 'only one nonzero value in ellipsis_mask is allowed'
+            ellipsis_start = i[0][0]
+            # since we don't expect values in begin, end values and take all range of values along ellipsis_start axis
+            begin_mask[ellipsis_start] = 0
+            end_mask[ellipsis_start] = 0
+
+            num = input_rank - len(begin) + np.count_nonzero(node.new_axis_mask[ellipsis_start:])
+            unroll_ellipsis = lambda mask, val=0: np.insert(mask, ellipsis_start + 1, [val] * num).astype(int)
+
+            new_axis_mask = unroll_ellipsis(new_axis_mask)
+            shrink_axis_mask = unroll_ellipsis(shrink_axis_mask)
+            begin_mask, end_mask = unroll_ellipsis(begin_mask), unroll_ellipsis(end_mask)
+            begin, end, strides = unroll_ellipsis(begin), unroll_ellipsis(end), unroll_ellipsis(strides, 1)
+
+        # from now slices are without ellipsis
+        slice_rank = len(begin)
+        slices = [[]] * slice_rank
+        in_idx = 0  # index along input tensor shapes, note that input_rank not necessary is equal to slice_rank
+        for i in range(slice_rank):
+            if new_axis_mask[i]:
+                slices[i] = np.newaxis
+            elif shrink_axis_mask[i]:
+                slices[i] = int(begin[i])
+                if slices[i] < 0:
+                    slices[i] += int(shape[in_idx])
```

Comment:
Looks like this line breaks reshape-ability of the model

Reply:
This logic does not affect shape calculation at all.

I agree that we loose some information in this line. But this whole thing with `slices` attribute (used by ConvertGroupedStridedSlice) is one big loose of information and `it not bug it's an essential feature`. For ConvertGroupedStridedSlice we need concrete values instead of "magic" numbers. 

E.g. `inp[:]` will be converted to concrete values `inp[0:dim_size]`. Firstly we meant 'take the whole range' but then 'take from 0 to dim_size'. This looks like breaking of reshape-ability but it is not, since this attribute is used only inside of MO where dim_size is just equal to shape upper bound. 

So I think that this line does not break reshape-ability. Or at least it breaks it to the extend to which CovertGroupedStrided slice breaks